static/app/utils/metrics/metricsContext.tsx

```diff
@@ -0,0 +1,105 @@
+import {createContext, useEffect, useState} from 'react';
+
+import {addErrorMessage} from 'sentry/actionCreators/indicator';
+import {t} from 'sentry/locale';
+import {MetricsMetaCollection, MetricsTagCollection, Organization} from 'sentry/types';
+import handleXhrErrorResponse from 'sentry/utils/handleXhrErrorResponse';
+import useApi from 'sentry/utils/useApi';
+
+interface ChildProps {
+  metas: MetricsMetaCollection;
+  tags: MetricsTagCollection;
+}
+
+export type MetricsContextProps = ChildProps | undefined;
+
+const MetricsContext = createContext<MetricsContextProps>(undefined);
+
+type ProviderProps = {
+  children: React.ReactNode | ((props: ChildProps) => React.ReactNode);
+  organization: Organization;
+  fields?: string[];
+  projects?: number[];
+  skipLoad?: boolean;
+};
+
+function MetricsProvider({
+  children,
+  projects,
+  fields,
+  organization,
+  skipLoad = false,
+}: ProviderProps) {
+  const api = useApi();
+  const [metas, setMetas] = useState<MetricsMetaCollection>({});
+  const [tags, setTags] = useState<MetricsTagCollection>({});
+
+  useEffect(() => {
+    let unmounted = false;
```

Comment:
I like what @k-fish proposed here in one of my PRs. Basically, since you have many of these, I would just store it once inside a ref.

Edit: Discard what I wrote above. This is actually not `unmounted` but more like `shouldCancelRequest`

Reply:
It's not just renaming, I suspect this specific `unmounted` doesn't do anything in practice, it's also why I was poking Jonas to change the var names in his, since with his it's related to stopping request-finished->setState mostly, not component mounting. The cleanup below isn't scoped such that the inflight requests would not setState after unmounted is modified.

Also, just a heads up, the way `useEffect` works with cleanups is _different when it has dependencies_. Dependencies changing cause cleanup to be called each time before useEffect runs again.